graphix/parameter.py

```diff
@@ -0,0 +1,311 @@
+        raise PlaceholderOperationError()
+
+    def sqrt(self) -> ExpressionOrNumber:
+        raise PlaceholderOperationError()
+
+    def __str__(self) -> str:
+        return f"{self.a} * {self.x} + {self.b}"
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, AffineExpression):
+            return self.a == other.a and self.x == other.x and self.b == other.b
+        return False
+
+    def subs(self, variable: Parameter, value: ExpressionOrNumber) -> ExpressionOrNumber:
+        if variable == self.x:
+            return self.a * value + self.b
+        return self
+
+
+class Placeholder(AffineExpression, Parameter):
+    """Placeholder for measurement angles.
+
+    These placeholder may appear in affine expressions.  Placeholders
+    and affine expressions may be used as angles in rotation gates of
+    :class:`Circuit` class or for the measurement angle of the
+    measurement commands.  Pattern optimizations such that
+    standardization, signal shifting and Pauli preprocessing can be
+    applied to patterns with placeholders.
+
+    These placeholders and affine expressions do not support arbitrary
+    computation and are not suitable for simulation.  Either use
+    :func:`Circuit.subs` or :func:`Pattern.subs` with an actual value
+    before the computation, or use a symbolic parameter
+    implementation, such as
+    https://github.com/TeamGraphix/graphix-symbolic .
+
+    """
+
+    def __init__(self, name: str) -> None:
+        """Create a new :class:`Placeholder` object.
+
+        Parameters
+        ----------
+        name : str
+            name of the parameter, used for binding values.
+        """
+        self.__name = name
+        super().__init__(a=1, x=self, b=0)
+
+    @property
+    def name(self) -> str:
+        return self.__name
+
+    def __str__(self) -> str:
+        return self.__name
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, Parameter):
+            return self is other
+        return super().__eq__(other)
+
+
+if sys.version_info >= (3, 10):
+    ExpressionOrFloat = Expression | float
+
+    ExpressionOrNumber = Expression | Number
+else:
+    ExpressionOrFloat = typing.Union[Expression, float]
+
+    ExpressionOrNumber = typing.Union[Expression, Number]
+
+
+def subs(value, variable: Parameter, substitute: ExpressionOrNumber):
+    """Generic substitution in `value`: if `value` implements the
+    method `subs`, then return `value.subs(variable, substitute)`
+    (coerced into a complex if the result is a number).  If `value`
+    does not implement `subs`, `value` is returned unchanged.
+
+    This function is used to apply substitution to collections where
+    some elements are Expression and other elements are just
+    plain numbers.
+
+    """
+    subs = getattr(value, "subs", None)
+    if subs:
+        new_value = subs(variable, substitute)
+        if isinstance(new_value, Number):
+            return complex(new_value)
```

Comment:
Seeing many type errors originating from `Number`.

Reply:
Fixed in https://github.com/TeamGraphix/graphix/commit/492b772ce7ec15e8d22566921a8fff6d87e99028.